rudder-core/src/main/scala/com/normation/rudder/services/reports/NodeChangesService.scala

```diff
@@ -33,102 +33,268 @@
+    val limit = since.getOrElse(DateTime.now.minusDays(changesMaxAge))
+    val now = DateTime.now
+
     // compute intervals
-    val intervals = new Interval(firstInterval,now) :: computeIntervals(firstInterval )
+    getInterval(limit, now)
+  }
+
 
+  /**
+   * Get all changes until a date specified as parameter and regroup them by interval of hours.
+   *
+   * Interval are 6 hour longs.
+   */
+  override def getChangesByInterval(since: Option[DateTime]) = {
 
     // Regroup changes by interval
     for {
-       changes <- getChanges(notBefore)
+       changes <- reportsRepository.getChangeReports(since.getOrElse(DateTime.now.minusDays(changesMaxAge)))
     } yield {
-      ( for {
-         interval <- intervals
-      } yield {
-        (interval,changes.filter(interval contains _.executionTimestamp))
-      } ).toMap
+      filterGroupByInterval(getCurrentValidIntervals(None), changes)
+    }
+  }
+}
+
+
+/**
+ * A cached version of NodeChangeService that:
+ * - get its config (interval lenght) and first results from an other service
+ * - is able to update a cache from other
+ */
+class CachedNodeChangesServiceImpl(
+    changeService: NodeChangesService
+) extends NodeChangesService with CachedRepository with Loggable {
+
+  private[this] var cache = Option.empty[ChangesByRule]
+
+  private[this] def cacheToLog(changes: ChangesByRule): String = {
+    changes.map { case(ruleId, x) =>
+      val byInt = x.map { case (int, ch) =>
+
+        val changeArray = ch.mkString("  ", "\n  ", "")
+
+        s" ${int}\n${changeArray}"
+      }.mkString(" ", "\n ", "")
+
+      s"${ruleId.value}\n${byInt}"
+    }.mkString("\n")
+  }
+
+  /**
+   * The only method authorised to directly look at
+   * the content of cache (other method can update the cache).
+   */
+  private[this] def initCache() : Box[ChangesByRule] = this.synchronized {
+    cache match {
+      case None =>
+        for {
+          changes <- changeService.getChangesByInterval(None)
+        } yield {
+          cache = Some(changes)
+          logger.debug("NodeChanges cache initialized")
+          logger.trace("NodeChanges cache content: " + cacheToLog(changes))
+          changes
+        }
+      case Some(changes) =>
+        logger.debug("NodeChanges cache hit")
+        logger.trace("NodeChanges cache content: " + cacheToLog(changes))
+        Full(changes)
+    }
+  }
+
+
+  /**
+   * For intervals "on", merge change1 and change2 (respective to rules). Intervals not in "on"
+   * are removed from the result.
+   * For a given rule, on a given interval, changes fomr change1 and change2 are added.
+   */
+  private[this] def merge(on: Seq[Interval], changes1: ChangesByRule, changes2: ChangesByRule): ChangesByRule = {
+    //shortcut that get map(k1)(k2) and return an empty seq if key not found
+    def get(changes: ChangesByRule, ruleId: RuleId, i: Interval): Seq[ResultRepairedReport] = {
+      changes.getOrElse(ruleId, Map()).getOrElse(i, Seq())
```

Comment:
oh, ok. So that's relevant then

So how are "old" interval cleaned? (when time goes from 11:59 to 12:00, how the three day old interval get removed ?)